{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "cZqtMk3n0m_y"
   },
   "source": [
    "### 1. Sandbox for Bioinformatics - NCBI APIs\n",
    "\n",
    "1. [NCBI/Entrez - PubMed ESearch (XML format)](#id1)\n",
    "2. [NCBI/Entrez - PubMed ESummary (XML format)](#id2)\n",
    "3. [NCBI/Entrez - Combined ESearch/ESummary (json format)](#id3)\n",
    "4. [NCBI/Entrez - EFetch Query and Parse using Biopython](#id4)\n",
    "5. [NCBI/Entrez - EPost as Class Object](#id5)\n",
    "6. [NCBI/Entrez - Entrez Functions](#id6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VWGhbzdCdD-k"
   },
   "source": [
    "#### Summary of Entrez Functions\n",
    "**esearch** - Searches and retrieves primary IDs (for use in EFetch, ELink,and ESummary) and term translations and optionally retains results for future use in the user's environment.\n",
    "\n",
    "**efetch** - Retrieves records in the requested format from a list of one or more primary IDs or from the user's environment.\n",
    "\n",
    "**esummary** - Retrieves document summaries from a list of primary IDs or from the user's environment.\n",
    "\n",
    "**einfo** - Provides field index term counts, last update, and available links for each database.\n",
    "\n",
    "**elink** - Checks for the existence of an external or Related Articles link from a list of one or more primary IDs. Retrieves primary IDs and relevancy scores for links to Entrez databases or Related Articles;  creates a hyperlink to the primary LinkOut provider for a specific ID and database, or lists LinkOut URLs and Attributes for multiple IDs.\n",
    "\n",
    "**epost** - Posts a file containing a list of primary IDs for future use in the user's environment to use with subsequent search strategies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 99,
     "status": "ok",
     "timestamp": 1735136383894,
     "user": {
      "displayName": "Gerald McCollam",
      "userId": "03782028358737990651"
     },
     "user_tz": 360
    },
    "id": "ojjpDZUxN8dE",
    "outputId": "08b005b6-53d9-4379-8c3e-767c71b5ec31"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# If Biopython is installed, pass, if not, install it.\n",
    "try:\n",
    "  import Bio\n",
    "except ImportError:\n",
    "  !pip install biopython\n",
    "\n",
    "try:\n",
    "  from dotenv import load_dotenv, dotenv_values\n",
    "except ImportError:\n",
    "  !pip install python-dotenv\n",
    "\n",
    "import os\n",
    "import Bio\n",
    "from dotenv import load_dotenv, dotenv_values\n",
    "load_dotenv() # get env vars\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "R4i5bm3-1b7O"
   },
   "source": [
    "<a name=\"id1\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 688,
     "status": "ok",
     "timestamp": 1734969630810,
     "user": {
      "displayName": "Gerald McCollam",
      "userId": "03782028358737990651"
     },
     "user_tz": 360
    },
    "id": "gEnwo_4WvbUM",
    "outputId": "16433824-6f14-496b-bc28-73ad19c923f9"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PMIDs retrieved:\n",
      "\n",
      "39623476\n",
      "39596027\n",
      "39588573\n",
      "\n",
      "The WebEnv value: MCID_6769891ef94c252a000227bc\n",
      "\n",
      "The QueryKey value: 1\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from lxml import etree\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# NCBI/Entrez - PubMed ESearch (XML format)\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "# API key\n",
    "api_key = os.getenv(\"NCBI_key\")\n",
    "\n",
    "# Base request\n",
    "base_entrez = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/'\n",
    "\n",
    "db = 'pubmed' # one of numerous others\n",
    "query = 'cancer[mesh]+AND+epigenomics[mesh]+AND+2024[pdat]'\n",
    "\n",
    "# 'query' is embedded in 'esearch'.\n",
    "esearch = (f'esearch.fcgi?db={db}&term={query}&usehistory=y&retmax=3')\n",
    "url = base_entrez + esearch # final url\n",
    "\n",
    "response = requests.get(url) # GET\n",
    "\n",
    "if response.status_code == 200: # successful\n",
    "  # Parse XML w/ lxml\n",
    "  tree = etree.fromstring(response.content)\n",
    "\n",
    "  # Use lxml's API to extract data\n",
    "  # (i.e. find all elements with a specific tag)\n",
    "  pmid_elements = tree.findall(\".//Id\")\n",
    "  webenv_element = tree.findall(\".//WebEnv\") # need this\n",
    "  querykey_element = tree.findall(\".//QueryKey\") # also need this\n",
    "  webenv = webenv_element[0].text\n",
    "  querykey = querykey_element[0].text\n",
    "\n",
    "  # Iterate over the PMID elements\n",
    "  print(f\"PMIDs retrieved:\\n\")\n",
    "  for element in pmid_elements: # PubMed Journal Article ID (PMID)\n",
    "      print(element.text)\n",
    "  print(f\"\\nThe WebEnv value: {webenv}\")\n",
    "  print(f\"\\nThe QueryKey value: {querykey}\")\n",
    "else:\n",
    "    print(\"Error fetching the XML data:\", response.status_code) # not successful"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oHf6Is9v_DYC"
   },
   "source": [
    "<a name=\"id2\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 226,
     "status": "ok",
     "timestamp": 1734969634941,
     "user": {
      "displayName": "Gerald McCollam",
      "userId": "03782028358737990651"
     },
     "user_tz": 360
    },
    "id": "F9oBNb5k6oCE",
    "outputId": "00159ce7-12e5-4b66-e801-28e39a82e959"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The URL retrieved: \n",
      "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&query_key=1&WebEnv=MCID_6769891ef94c252a000227bc&retmax=1\n",
      "\n",
      "The PMID: 39623476\n",
      "\n",
      "DOCUMENT TREE:\n",
      "\n",
      "Parent: DocSum\n",
      "  Child: Id\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "    Grandchild: Item, Attribute: {'Name': 'Author', 'Type': 'String'}, Text: Kelly K\n",
      "    Grandchild: Item, Attribute: {'Name': 'Author', 'Type': 'String'}, Text: Scherer M\n",
      "    Grandchild: Item, Attribute: {'Name': 'Author', 'Type': 'String'}, Text: Braun MM\n",
      "    Grandchild: Item, Attribute: {'Name': 'Author', 'Type': 'String'}, Text: Lutsik P\n",
      "    Grandchild: Item, Attribute: {'Name': 'Author', 'Type': 'String'}, Text: Plass C\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "    Grandchild: Item, Attribute: {'Name': 'Lang', 'Type': 'String'}, Text: English\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "    Grandchild: Item, Attribute: {'Name': 'PubType', 'Type': 'String'}, Text: Journal Article\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "    Grandchild: Item, Attribute: {'Name': 'pubmed', 'Type': 'String'}, Text: 39623476\n",
      "    Grandchild: Item, Attribute: {'Name': 'pmc', 'Type': 'String'}, Text: PMC11613708\n",
      "    Grandchild: Item, Attribute: {'Name': 'pmcid', 'Type': 'String'}, Text: pmc-id: PMC11613708;\n",
      "    Grandchild: Item, Attribute: {'Name': 'doi', 'Type': 'String'}, Text: 10.1186/s13059-024-03446-w\n",
      "    Grandchild: Item, Attribute: {'Name': 'pii', 'Type': 'String'}, Text: 10.1186/s13059-024-03446-w\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "    Grandchild: Item, Attribute: {'Name': 'received', 'Type': 'Date'}, Text: 2024/04/24 00:00\n",
      "    Grandchild: Item, Attribute: {'Name': 'accepted', 'Type': 'Date'}, Text: 2024/11/26 00:00\n",
      "    Grandchild: Item, Attribute: {'Name': 'medline', 'Type': 'Date'}, Text: 2024/12/03 06:25\n",
      "    Grandchild: Item, Attribute: {'Name': 'pubmed', 'Type': 'Date'}, Text: 2024/12/03 00:23\n",
      "    Grandchild: Item, Attribute: {'Name': 'entrez', 'Type': 'Date'}, Text: 2024/12/02 23:55\n",
      "    Grandchild: Item, Attribute: {'Name': 'pmc-release', 'Type': 'Date'}, Text: 2024/12/03 00:00\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n",
      "  Child: Item\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from lxml import etree\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# NCBI/Entrez - PubMed ESummary (XML format)\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "# API key\n",
    "api_key = os.getenv(\"NCBI_key\")\n",
    "\n",
    "# Build request\n",
    "base_entrez = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/\"\n",
    "\n",
    "db = 'pubmed' # one of numerous databases maintained by NCBI\n",
    "\n",
    "# Construct query to retrieve a single PMID only\n",
    "esummary = (f\"esummary.fcgi?db={db}&query_key={querykey}&WebEnv={webenv}&retmax=1\")\n",
    "\n",
    "url = base_entrez + esummary\n",
    "\n",
    "response = requests.get(url) # GET\n",
    "print(f\"The URL retrieved: \\n{url}\")\n",
    "\n",
    "if response.status_code == 200: # success\n",
    "  # Parse XML w/ lxml\n",
    "  tree = etree.fromstring(response.content)\n",
    "  # Get the element PMID\n",
    "  for element in tree.xpath(\".//Id\"):\n",
    "    print(f\"\\nThe PMID: {element.text}\")\n",
    "\n",
    "  print(f\"\\nDOCUMENT TREE:\\n\")\n",
    "  # Navigate the document tree\n",
    "  for parent in tree:\n",
    "    print(f\"Parent: {parent.tag}\")\n",
    "    for child in parent:\n",
    "      print(f\"  Child: {child.tag}\")\n",
    "      for grandchild in child:\n",
    "        print(f\"    Grandchild: {grandchild.tag}, Attribute: {grandchild.attrib}, Text: {grandchild.text}\")\n",
    "else:\n",
    "    print(\"Error fetching the XML data:\", response.status_code) # not successful"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YLXwIcYRaNt3"
   },
   "source": [
    "<a name=\"id3\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 382,
     "status": "ok",
     "timestamp": 1734969640570,
     "user": {
      "displayName": "Gerald McCollam",
      "userId": "03782028358737990651"
     },
     "user_tz": 360
    },
    "id": "IX_xbkZWaQuw",
    "outputId": "9d200a62-bdfa-4b8b-a436-495094ab6749"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"header\": {\n",
      "    \"type\": \"esummary\",\n",
      "    \"version\": \"0.3\"\n",
      "  },\n",
      "  \"result\": {\n",
      "    \"uids\": [\n",
      "      \"2871752916\"\n",
      "    ],\n",
      "    \"2871752916\": {\n",
      "      \"uid\": \"2871752916\",\n",
      "      \"term\": \"2871752916\",\n",
      "      \"caption\": \"PQ779047\",\n",
      "      \"title\": \"Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/NY-CUIMC-NP-15699/2023 ORF1ab polyprotein (ORF1ab) and ORF1a polyprotein (ORF1ab) genes, partial cds; surface glycoprotein (S), ORF3a protein (ORF3a), envelope protein (E), membrane glycoprotein (M), ORF6 protein (ORF6), ORF7a protein (ORF7a), and ORF7b (ORF7b) genes, complete cds; ORF8 gene, complete sequence; and nucleocapsid phosphoprotein (N) and ORF10 protein (ORF10) genes, complete cds\",\n",
      "      \"extra\": \"gi|2871752916|gb|PQ779047.1|\",\n",
      "      \"gi\": 2871752916,\n",
      "      \"createdate\": \"2024/12/18\",\n",
      "      \"updatedate\": \"2024/12/18\",\n",
      "      \"flags\": \"\",\n",
      "      \"taxid\": 2697049,\n",
      "      \"slen\": 29177,\n",
      "      \"biomol\": \"genomic\",\n",
      "      \"moltype\": \"rna\",\n",
      "      \"topology\": \"not-set\",\n",
      "      \"sourcedb\": \"insd\",\n",
      "      \"segsetsize\": \"\",\n",
      "      \"projectid\": \"0\",\n",
      "      \"genome\": \"\",\n",
      "      \"subtype\": \"isolate|host|gb_acronym|country|isolation_source|collection_date\",\n",
      "      \"subname\": \"SARS-CoV-2/human/USA/NY-CUIMC-NP-15699/2023|Homo sapiens|SARS-CoV-2|USA: New York City|nasopharyngeal swab|2023-11-09\",\n",
      "      \"assemblygi\": \"\",\n",
      "      \"assemblyacc\": \"\",\n",
      "      \"tech\": \"\",\n",
      "      \"completeness\": \"\",\n",
      "      \"geneticcode\": \"1\",\n",
      "      \"strand\": \"\",\n",
      "      \"organism\": \"Severe acute respiratory syndrome coronavirus 2\",\n",
      "      \"strain\": \"\",\n",
      "      \"biosample\": \"\",\n",
      "      \"statistics\": \"\",\n",
      "      \"properties\": {\n",
      "        \"na\": \"1\",\n",
      "        \"value\": \"1\"\n",
      "      },\n",
      "      \"oslt\": {\n",
      "        \"indexed\": true,\n",
      "        \"value\": \"PQ779047.1\"\n",
      "      },\n",
      "      \"accessionversion\": \"PQ779047.1\"\n",
      "    }\n",
      "  }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# Combined ESearch/ESummary (json format)\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "# API key for NCBI\n",
    "api_key = os.getenv(\"NCBI_key\")\n",
    "\n",
    "# Base request\n",
    "base_entrez = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/'\n",
    "\n",
    "db = 'nucleotide'\n",
    "query = 'viruses[orgn]'\n",
    "\n",
    "# retmode is `json`.\n",
    "esearch = (f'esearch.fcgi?db={db}&term={query}&usehistory=y&retmode=json&retmax=1')\n",
    "url = base_entrez + esearch # final url\n",
    "\n",
    "response = requests.get(url) # STEP 1: Search ----------------------------------\n",
    "\n",
    "if response.status_code == 200:\n",
    "  search_data = response.json() # returns a Python dict\n",
    "  webenv = search_data['esearchresult']['webenv'] # need this\n",
    "  querykey = search_data['esearchresult']['querykey'] # and this\n",
    "\n",
    "# Construct query to retrieve a single summary, as json.\n",
    "esummary = (f\"esummary.fcgi?db={db}&query_key={querykey}&WebEnv={webenv}&retmode=json&retmax=1\")\n",
    "\n",
    "url = base_entrez + esummary\n",
    "\n",
    "response = requests.get(url) # STEP 2: Retrieve --------------------------------\n",
    "\n",
    "if response.status_code == 200:\n",
    "  summary_data = response.json() # return a Python dict\n",
    "\n",
    "# Convert to JSON formatted string with 2 spaces of indentation\n",
    "json_str = json.dumps(summary_data, indent=2)\n",
    "\n",
    "print(json_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "aPVn4A2GZKYq"
   },
   "source": [
    "<a name=\"id4\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 547,
     "status": "ok",
     "timestamp": 1734969647125,
     "user": {
      "displayName": "Gerald McCollam",
      "userId": "03782028358737990651"
     },
     "user_tz": 360
    },
    "id": "Iv0OhnTOD2Ze",
    "outputId": "82821fb6-b7c5-4f41-d1da-850e88ae334e"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record id: PQ779047.1\n",
      "Record description: Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/NY-CUIMC-NP-15699/2023 ORF1ab polyprotein (ORF1ab) and ORF1a polyprotein (ORF1ab) genes, partial cds; surface glycoprotein (S), ORF3a protein (ORF3a), envelope protein (E), membrane glycoprotein (M), ORF6 protein (ORF6), ORF7a protein (ORF7a), and ORF7b (ORF7b) genes, complete cds; ORF8 gene, complete sequence; and nucleocapsid phosphoprotein (N) and ORF10 protein (ORF10) genes, complete cds\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# NCBI/Entrez - EFetch using Biopython\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "def download_file(url):\n",
    "  \"\"\"Downloads a SINGLE file given an NCBI eFetch URL.\n",
    "     DOES NOT HANDLE MORE THAN ONE FILE!\n",
    "  \"\"\"\n",
    "  with open('sequence.gb', 'wb') as out_file:\n",
    "    content = requests.get(url, stream=True).content\n",
    "    out_file.write(content)\n",
    "\n",
    "def seqrecord_to_dict(record):\n",
    "  \"\"\"Converts a Bio.SeqRecord.SeqRecord to a plain Python dictionary.\n",
    "  \"\"\"\n",
    "  return {\n",
    "      \"id\": record.id,\n",
    "      \"name\": record.name,\n",
    "      \"description\": record.description,\n",
    "      \"seq\": str(record.seq),\n",
    "      \"features\": [feature.__dict__ for feature in record.features],\n",
    "      \"annotations\": record.annotations,\n",
    "      \"letter_annotations\": record.letter_annotations\n",
    "  }\n",
    "\n",
    "# API key for NCBI\n",
    "api_key = os.getenv(\"NCBI_key\")\n",
    "\n",
    "db = 'nucleotide'\n",
    "query = 'viruses[orgn]'\n",
    "\n",
    "# Base request\n",
    "base_entrez = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/'\n",
    "\n",
    "# Construct query to retrieve a single summary, as json.\n",
    "efetch = (f\"efetch.fcgi?db={db}&query_key={querykey}&WebEnv={webenv}&rettype=abstract&retmode=text&retmax=1\")\n",
    "\n",
    "url = base_entrez + efetch\n",
    "\n",
    "download_file(url)\n",
    "\n",
    "with open(\"sequence.gb\", \"r\") as handle:\n",
    "  record = SeqIO.read(handle, \"genbank\")\n",
    "\n",
    "print(f\"Record id: {record.id}\")\n",
    "print(f\"Record description: {record.description}\")\n",
    "\n",
    "result = seqrecord_to_dict(record)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "IgeUoKy8_Tid"
   },
   "source": [
    "<a name=\"id5\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QK9JMs3G_Vd2"
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# NCBI/Entrez - EPost as Class Objects - First Try\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "class Parameter(object):\n",
    "\n",
    "  db = {1: 'assembly', 2: 'clinvar', 3: 'dbvar', 4: 'gap', \\\n",
    "        5: 'gene', 6: 'genome', 7: 'mesh', 8: 'nuccore', 9: 'nucleotide', \\\n",
    "       10: 'omim', 11: 'protein', 12: 'pubmed', 13: 'snp', 14: 'sra'}\n",
    "\n",
    "  uids = [0]\n",
    "  webenv = None\n",
    "  retmode = 'json'\n",
    "  rettype = None\n",
    "  querykey = None\n",
    "\n",
    "\n",
    "class EutilsRequest(object):\n",
    "  def __init__(self, db):\n",
    "    self.db = db\n",
    "    self.id = None\n",
    "    self.querykey = None\n",
    "    self.tool = 'gene_genie'\n",
    "    self.url = None\n",
    "    self.contact = 'gerald.mccollam@gmail.com'\n",
    "    self.apikey = ''\n",
    "    self.status = 2\n",
    "    self.request_error = None\n",
    "    self.size = 1\n",
    "\n",
    "  def prepare_base_qry(self, extend=None):\n",
    "    \"\"\"\n",
    "    Returns instance attributes required for every request.\n",
    "\n",
    "    \"\"\"\n",
    "    if self.apikey:\n",
    "      {'email' : self.contact, 'tool' : self.tool, 'db' : self.db}.update({'api_key' : self.apikey})\n",
    "    if extend:\n",
    "      {'email' : self.contact, 'tool' : self.tool, 'db' : self.db}.update(extend)\n",
    "    return {'email' : self.contact, 'tool' : self.tool, 'db' : self.db}\n",
    "\n",
    "class EpostRequest(EutilsRequest):\n",
    "  def __init__(self, parameter):\n",
    "    super().__init__(parameter.db)\n",
    "    self.uids = parameter.uids\n",
    "    self.size = len(parameter.uids)\n",
    "    self.webenv = parameter.webenv\n",
    "    self.querykey = parameter.querykey\n",
    "    self.retmode = parameter.retmode\n",
    "    self.rettype = parameter.retmode\n",
    "\n",
    "\n",
    "  def get_post_parameter(self):\n",
    "    return self.prepare_base_qry(extend={'id':','.join(str(x) for x in self.uids),\n",
    "                                         'WebEnv':self.webenv})\n",
    "\n",
    "  def dump(self):\n",
    "    return {'db' : self.db,\n",
    "            'WebEnv':self.webenv,\n",
    "            'query_key' : self.querykey,\n",
    "            'uids' : self.uids,\n",
    "            'retmode' : self.retmode,\n",
    "            'rettype' : self.rettype,\n",
    "            'retstart' : self.retstart,\n",
    "            'retmax' : self.retmax,\n",
    "            'request_size' : self.reqsize\n",
    "            }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0YnUKi6XjmFb"
   },
   "source": [
    "<a name=\"id6\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "iQioK5lbiB_h"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import warnings\n",
    "import json\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "from urllib.parse import urlencode\n",
    "from dotenv import load_dotenv, dotenv_values \n",
    "\n",
    "# ------------------------------------------------------------------------------\n",
    "# NCBI/Entrez - Entrez functions based on Biopython but using the Requests lib.\n",
    "# NCBI defaults to XML. Try to use json wherever possible.\n",
    "# ------------------------------------------------------------------------------\n",
    "\n",
    "load_dotenv() # get env vars\n",
    "\n",
    "email = 'gerald.mccollam@gmail.com'\n",
    "max_tries = 3\n",
    "sleep_between_tries = 15\n",
    "tool = \"gene_genie\"\n",
    "api_key = os.getenv(\"NCBI_key\")\n",
    "local_cache = None\n",
    "\n",
    "def esearch(db, term, **keywords):\n",
    "  '''This function searches and retrieves primary IDs (for use in EFetch,\n",
    "     ELink and ESummary).\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\"\n",
    "  variables = {\"db\": db, \"term\": term}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response\n",
    "\n",
    "def efetch(db, **keywords):\n",
    "  '''This function retrieves records in the requested format from a list or\n",
    "     set of one or more UIs or from user's environment.\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "  variables = {\"db\": db}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response\n",
    "\n",
    "def esummary(**keywords):\n",
    "  '''This function retrieves document summaries from a list of primary IDs or\n",
    "     from the user's environment.\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\"\n",
    "  variables = {}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  #print(request)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response\n",
    "\n",
    "def einfo(**keywords):\n",
    "  '''This function returns a summary of the Entrez databases as a results handle.\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\"\n",
    "  variables = {}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  print(request)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response\n",
    "\n",
    "def elink(**keywords):\n",
    "  '''This function checks for the existence of an external or Related Articles\n",
    "     link from a list of one or more primary IDs;  retrieves IDs and relevancy\n",
    "     scores for links to Entrez databases or Related Articles.\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\"\n",
    "  variables = {}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response\n",
    "\n",
    "def epost(db, **keywords):\n",
    "  '''Posts a file containing a list of primary IDs for future use.\n",
    "  '''\n",
    "  base_cgi = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi\"\n",
    "  variables = {\"db\": db}\n",
    "  variables.update(keywords)\n",
    "  request = _construct(base_cgi, variables)\n",
    "  try:\n",
    "    response = requests.get(base_cgi, request)  \n",
    "    response.raise_for_status()\n",
    "  except requests.exceptions.RequestException as e:\n",
    "    print(\"Request failed:\", str(e))\n",
    "  return response \n",
    "    \n",
    "def _get_params(params, join_ids=True):\n",
    "  pass\n",
    "\n",
    "def _format_ids(ids):\n",
    "    '''\n",
    "    Input: a single ID (int or str), or iterable of strings/ints,\n",
    "    or a string of IDs separated by commas.\n",
    "    '''\n",
    "    if isinstance(ids, int):\n",
    "        return str(ids) # Single int, convert it to str.\n",
    "\n",
    "    if isinstance(ids, str):\n",
    "        # Multiple IDs, remove white space.\n",
    "        return \",\".join(id.strip() for id in ids.split(\",\"))\n",
    "\n",
    "    # Not a string or integer, assume iterable\n",
    "    return \",\".join(map(str, ids))\n",
    "\n",
    "def _has_api_key(request):\n",
    "  return \"api_key=\" in request.full_url\n",
    "\n",
    "def _construct(base_cgi, params=None, join_ids=True):\n",
    "  '''\n",
    "  '''\n",
    "  if params is None:\n",
    "    params = {}\n",
    "  params.setdefault(\"tool\", tool)\n",
    "  params.setdefault(\"email\", email)\n",
    "  params.setdefault(\"api_key\", api_key)\n",
    "  params.setdefault(\"usehistory\", 'y')\n",
    "  params.setdefault(\"retmode\", 'json')\n",
    "\n",
    "  if join_ids and \"id\" in params:\n",
    "      params[\"id\"] = _format_ids(params[\"id\"])\n",
    "\n",
    "  return params\n",
    "\n",
    "def get_DOI_urls(resp):\n",
    "  '''Take a response object as input, return a list of qualified DOI url[s].\n",
    "  '''\n",
    "  DOI_URL = 'https://doi.org/'\n",
    "\n",
    "  pmids = resp.json()['result']['uids']\n",
    "  pmid_list = []\n",
    "  if len(pmids) == 0:\n",
    "    pass\n",
    "  if len(pmids) == 1:\n",
    "    elocationid =  resp.json()['result'][pmids[0]]['elocationid']\n",
    "    pmid_list.append(DOI_URL + elocationid.split(\" \")[1])\n",
    "  elif len(pmids) > 1:\n",
    "    for pmid in pmids:\n",
    "      elocationid =  resp.json()['result'][pmid]['elocationid']\n",
    "      pmid_list.append(DOI_URL + elocationid.split(\" \")[1])\n",
    "  else:\n",
    "    pass\n",
    "\n",
    "  return pmid_list\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "  resp = esearch(db=\"pubmed\", retmax=5, term=\"cancer[mesh] epigenomics[mesh] 2024[pdat]\")\n",
    "  querykey = resp.json()['esearchresult']['querykey']\n",
    "  webenv = resp.json()['esearchresult']['webenv']\n",
    "  resp_summary = esummary(db=\"pubmed\", retmax=2, query_key=querykey, webenv=webenv)\n",
    "  urls = get_DOI_urls(resp_summary)\n",
    "  print(urls)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyOZ9QJ4pk4SuzSakHNa6Gbl",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
